### PR TITLE
fix(chain muxer): do not assemble gossipsub blocks into full tipsets during bootstrap

### PR DESCRIPTION
**Pull request by hanabi1224**
## Summary of changes

As a follow-up of https://github.com/ChainSafe/forest/pull/3713
It should be unnecessary to make bitswap requests during BOOTSTRAP. This PR tries to avoid that and improve BOOTSTRAP performance 

Changes introduced in this pull request:

- Do not assemble gossipsub blocks into full tipsets during bootstrap, to avoid making unnecessary underlying bitswap requests
- Do not assemble gossipsub blocks into full tipsets when they are older than `head_epoch - chain_finality`, to reduce unnecessary underlying bitswap requests
- Do not assemble hello request into full tipsets when they are older than `head_epoch - chain_finality`, to reduce unnecessary underlying chain exchange requests

Manually tested on mainnet, forest is able to sync with normal performance.

## Reference issue to close (if applicable)

Closes

## Change checklist

- [x] I have performed a self-review of my own code,
- [x] I have made corresponding changes to the documentation,
- [x] I have added tests that prove my fix is effective or that my feature works (if possible),
- [x] I have made sure the [CHANGELOG][1] is up-to-date. All user-facing changes should be reflected in this document.

[1]: https://github.com/ChainSafe/forest/blob/main/CHANGELOG.md

### CODE REVIEW

File: src/chain_sync/chain_muxer.rs [hunk 1]
```diff
@@ -365,28 +370,48 @@ where
...
-                    Ok(tipset) => tipset,
-                    Err(why) => {
-                        debug!("Querying full tipset failed: {}", why);
-                        return Err(why);
+                // Retrieve tipset from store based on passed in TipsetKeys
+                if let Ok(Some(tipset)) = chain_store.load_tipset(&tipset_keys) {
+                    // Skip old tipsets
+                    if chain_store.is_epoch_finalized(tipset.epoch()) {
+                        return Ok(None);
+                    } else {
```
Comment (ruseinov) [inline src/chain_sync/chain_muxer.rs:395]: nit: else is redundant here.

File: src/chain_sync/chain_muxer.rs [hunk 2]
```diff
@@ -365,28 +370,48 @@ where
...
+                        )
+                        .await
+                        {
+                            Ok(tipset) => tipset,
+                            Err(why) => {
+                                debug!("Querying full tipset failed: {}", why);
+                                return Err(why);
+                            }
+                        };
+                        (full_tipset, source)
```
Comment (ruseinov) [inline src/chain_sync/chain_muxer.rs:410]: nit: if you return here - you can get rid of the `else` block below and just do a return there too.

File: src/chain_sync/chain_muxer.rs [hunk 3]
```diff
@@ -436,9 +461,24 @@ where
                         .with_label_values(&[metrics::values::PUBSUB_BLOCK])
                         .inc();
                     // Assemble full tipset from block
-                    let tipset =
-                        Self::gossipsub_block_to_full_tipset(b, source, network.clone()).await?;
-                    (tipset, source)
+                    if let GossipBlockProcessingStrategy::GetFullTipset =
```
Comment (ruseinov) [inline src/chain_sync/chain_muxer.rs:464]: nit: same here, there could be less nesting if we get rid of the `else` blocks.

File: src/chain_sync/chain_muxer.rs [hunk 4]
```diff
@@ -517,7 +550,15 @@ where
             .with_label_values(&[source.to_string().as_str()])
             .set(tipset.epoch());
 
-        Ok(Some((tipset, source)))
+        if should_skip {
```
Comment (ruseinov) [inline src/chain_sync/chain_muxer.rs:553]: nit: unnecessary `else` block.

**Comment by ruseinov** [comment]
Some minor readability nits. It would also be nice to see some tests for this if at all possible.

File: src/chain_sync/chain_muxer.rs (hunk 4 above)
Comment (hanabi1224) [reply src/chain_sync/chain_muxer.rs:553]: Fixed

File: src/chain_sync/chain_muxer.rs (hunk 3 above)
Comment (hanabi1224) [reply src/chain_sync/chain_muxer.rs:464]: Fixed.

File: src/chain_sync/chain_muxer.rs (hunk 2 above)
Comment (hanabi1224) [reply src/chain_sync/chain_muxer.rs:410]: The `full_tipset` is assigned to a variable and processed later, early return here breaks the logic.

File: src/chain_sync/chain_muxer.rs (hunk 1 above)
Comment (hanabi1224) [reply src/chain_sync/chain_muxer.rs:395]: Fixed.

**Comment by hanabi1224** [comment]
> Some minor readability nits. It would also be nice to see some tests for this if at all possible.

@ruseinov This should have been covered by the integration test scripts, and I've also manually tested on mainnet

**Comment by hanabi1224** [comment]
Kindly ping.  @lemmih any feedback on this one?

**Comment by lemmih** [comment]
> Kindly ping. @lemmih any feedback on this one?

I'm unsure what this PR is attempting to do. Is it purely a performance optimization? If so, how much does it improve performance? Does it justify the extra code complexity?

**Comment by hanabi1224** [comment]
> I'm unsure what this PR is attempting to do. Is it purely a performance optimization? If so, how much does it improve performance? Does it justify the extra code complexity?

@lemmih Besides performance optimization, I think it makes the logic of chain muxer more correct and clean. When a forest node is bootstrapped with a snapshot that is a few thousand epochs behind the network, it takes hours to sync(which is the BOOTSTRAP stage). During such a long BOOTSTRAP, a forest node makes 10k-100k bitswap requests(from my observation) to assemble the gossip blocks(passively received from the p2p network) into full tipsets and store the messages into its database. Neither the network nor the database activities seem to be necessary. Disabling the gossip-block-to-full-tipset assembly saves network bandwidth and disk space. Does that make sense?

**Comment by LesnyRumcajs** [comment]
@hanabi1224 @lemmih what do we do with this PR?

**Comment by lemmih** [comment]
> @hanabi1224 @lemmih what do we do with this PR?

I'm leaning toward closing this PR.

**Pull request closed without merging**